src/metabase/db/setup.clj

```diff
@@ -120,6 +124,33 @@
       (log/infof "Successfully verified %s %s application database connection. %s"
                  (.getDatabaseProductName metadata) (.getDatabaseProductVersion metadata) (u/emoji "✅")))))
 
+(mu/defn- check-encryption
+  "Ensure encryption env variable is correctly set if needed, and encrypt the database if it needs to be
+  Encryption status is tracked by an 'encryption-check' value in the settings table.
+  NOTE: the encryption-check setting is not managed like most settings with 'defsetting' so we can manage checking the raw values in the database"
+  [db-type :- :keyword
+   data-source :- (ms/InstanceOfClass javax.sql.DataSource)]
+  (when (or (liquibase/table-exists? "setting" (.getConnection ^javax.sql.DataSource data-source)) (liquibase/table-exists? "SETTING" (.getConnection ^javax.sql.DataSource data-source)))
+    (try
+      (t2/insert! :conn data-source :setting {:key "encryption-check" :value "unencrypted"})
+      (catch Exception e (log/debug "encryption-check already exists" (.getMessage e))))
+    (let [raw (-> (t2/select-one :conn data-source :setting :key "encryption-check")
+                  :value)
+          looks-encrypted (not= raw "unencrypted")]
+      (log/debug "Checking encryption configuration")
+      (if looks-encrypted
+        (if (encryption/default-encryption-enabled?)
+          (if (string/valid-uuid? (encryption/maybe-decrypt raw))
+            (log/debug "Database encrypted and MB_ENCRYPTION_SECRET_KEY correctly configured")
+            (throw (Exception. "Database was encrypted with a different key than the MB_ENCRYPTION_SECRET_KEY environment contains")))
+          (throw (Exception. "Database is encrypted but the MB_ENCRYPTION_SECRET_KEY environment variable was NOT set")))
+        (if (encryption/default-encryption-enabled?)
+          (do
+            (log/info "New MB_ENCRYPTION_SECRET_KEY environment variable set. Encrypting database...")
+            (mdb.encryption/encrypt-db db-type data-source nil)
+            (log/info "Database encrypted..." (u/emoji "✅")))
+          (log/debug "Database not encrypted and MB_ENCRYPTION_SECRET_KEY env variable not set."))))))
```

Comment:
Usually nested ifs like this is kind-of a code smell.

Since we don't need to _return_ anything when we throw, you could keep the nesting down by using when-not and throwing, something like:

```
(if looks-encrypted
  (do 
    (when-not (encryption/default-encryption-enabled?)
      (throw (Exception. "Database is encrypted but the MB_ENCRYPTION_SECRET_KEY environment variable was NOT set")))
    (when-not (string/valid-uuid? (encryption/maybe-decrypt raw))
      (throw (Exception. "Database was encrypted with a different key than the MB_ENCRYPTION_SECRET_KEY environment contains")))
    (log/debug "Database encrypted and MB_ENCRYPTION_SECRET_KEY correctly configured"))

  (do
    (when-not (encryption/default-encryption-enabled?)
      (log/debug "Database not encrypted and MB_ENCRYPTION_SECRET_KEY env variable not set."))
    (log/info "New MB_ENCRYPTION_SECRET_KEY environment variable set. Encrypting database...")
    (mdb.encryption/encrypt-db db-type data-source nil)
    (log/info "Database encrypted..." (u/emoji "✅"))))
```

----

Also, this is a preference thing: but I'd prefer throwing `ex-info` with (or without) some data in the map to describe what's happening to `Exception.`.